starport/pkg/spn/chain.go

```diff
@@ -2,13 +2,129 @@ package spn
 
 import (
 	"context"
-	"github.com/tendermint/starport/starport/pkg/jsondoc"
+	"golang.org/x/sync/errgroup"
 	"time"
 
 	"github.com/cosmos/cosmos-sdk/types"
 	genesistypes "github.com/tendermint/spn/x/genesis/types"
+	"github.com/tendermint/starport/starport/pkg/jsondoc"
 )
 
+// ChainSummary represents the summary of a chain in Genesis module of SPN.
+type ChainSummary struct {
+	ChainID            string
+	Source             string
+	TotalValidators    int
+	ApprovedValidators int
+	TotalProposals     int
+	ApprovedProposals  int
+}
+
+// ChainList lists chain summaries
+func (c *Client) ChainList(ctx context.Context, accountName string) ([]ChainSummary, error) {
+	clientCtx, err := c.buildClientCtx(accountName)
+	if err != nil {
+		return []ChainSummary{}, err
+	}
+
+	// List the chains
+	q := genesistypes.NewQueryClient(clientCtx)
+	req := &genesistypes.QueryListChainsRequest{}
+	chainList, err := q.ListChains(ctx, req)
+	if err != nil {
+		return []ChainSummary{}, err
+	}
+
+	// Get the summary of each chain
+	chainSummaries := make([]ChainSummary, len(chainList.Chains))
+	chainSummariesGroup := new(errgroup.Group)
+
+	for i, chain := range chainList.Chains {
+		i, chain := i, chain // https://golang.org/doc/faq#closures_and_goroutines
+		chainSummariesGroup.Go(func() error {
+			chainSummaryGroup := new(errgroup.Group)
+
+			var chainSummary ChainSummary
+			chainSummary.ChainID = chain.ChainID
+			chainSummary.Source = chain.SourceURL
+
+			// Get the number of validators
+			chainSummaryGroup.Go(func() error {
+				reqValidators := &genesistypes.QueryListProposalsRequest{
+					ChainID: chain.ChainID,
+					Status:  genesistypes.ProposalStatus_ANY_STATUS,
+					Type:    genesistypes.ProposalType_ADD_VALIDATOR,
+				}
+				resValidators, err := q.ListProposals(ctx, reqValidators)
+				if err != nil {
+					return err
+				}
+				chainSummary.TotalValidators = len(resValidators.Proposals)
```

Comment:
This and others modify `chainSummary` will cause to a data race. We need to protect it with a mutex.

Reply:
after a second check, no problems with this one :+1: